app/scripts/migrations/104.ts

```diff
@@ -0,0 +1,117 @@
+import {
+  EthAccountType,
+  InternalAccount,
+  EthMethod,
+} from '@metamask/keyring-api';
+import { sha256FromString } from 'ethereumjs-util';
+import { v4 as uuid } from 'uuid';
+import { cloneDeep } from 'lodash';
+
+type VersionedData = {
+  meta: { version: number };
+  data: Record<string, unknown>;
+};
+
+interface Identity {
+  name: string;
+  address: string;
+  lastSelected?: number;
+}
+
+export const version = 104;
+
+/**
+ * This migration does the following:
+ *
+ * - Creates a default state for AccountsController.
+ * - Copies identities and selectedAddress from the PreferencesController to
+ * the AccountsController state as internal accounts and selectedAccount.
+ *
+ * @param originalVersionedData - Versioned MetaMask extension state, exactly what we persist to dist.
+ * @param originalVersionedData.meta - State metadata.
+ * @param originalVersionedData.meta.version - The current state version.
+ * @param originalVersionedData.data - The persisted MetaMask state, keyed by controller.
+ * @returns Updated versioned MetaMask extension state.
+ */
+export async function migrate(
+  originalVersionedData: VersionedData,
+): Promise<VersionedData> {
+  const versionedData = cloneDeep(originalVersionedData);
+  versionedData.meta.version = version;
+  migrateData(versionedData.data);
+  return versionedData;
+}
+
+function migrateData(state: Record<string, unknown>): void {
+  createDefaultAccountsController(state);
+  createInternalAccountsForAccountsController(state);
+  createSelectedAccountForAccountsController(state);
+}
+
+function createDefaultAccountsController(state: Record<string, any>) {
+  state.AccountsController = {
+    internalAccounts: {
+      accounts: {},
+      selectedAccount: '',
+    },
+  };
+}
+
+function createInternalAccountsForAccountsController(
+  state: Record<string, any>,
+) {
+  const identities: {
+    [key: string]: Identity;
+  } = state.PreferencesController?.identities || {};
+
+  if (Object.keys(identities).length === 0) {
+    return;
+  }
+
+  const accounts: Record<string, InternalAccount> = {};
```

Comment:
What if this property somehow exists already? Wondering if we want to do something like

``` suggestion
  const accounts: Record<string, InternalAccount> =
    cloneDeep(state.AccountsController.internalAccounts.accounts);

Reply:
Similarly to the previous comment, the property is being introduced for the first time in this PR